### ISSUE TITLE
Removing `ptr` and adding `ref`

### ISSUE THREAD

**Issue by dj2** [gpuweb/gpuweb#1334]
I'd like to propose we remove the `ptr` type from WGSL. I've found this type to be especially confusing because we also have things like `var` which is a pointer to some bit of member, but it's different from `ptr`.

In the place of `ptr` I propose we add a `ref` keyword which is used similar to `var`.

```groovy
fn main() -> void {
  var a : i32 = 2;
  ref b : i32 = a;   // Take a reference to a

  const c : ptr<function, i32> = a;  // current ptr syntax
}
```

```groovy
var <private> a : i32 = 2;
fn main() -> void {
  ref b : i32 = a;   // b inherits the `private` storage class from a
  ref<private> c : i32 = a;  // Even if provided, the storage class must match.

  const b : ptr<private, i32> = a;  // current ptr syntax
}
```

When using a ref as a parameter to a function, the storage class is required.
```groovy
fn b(ref<private> a : i32) -> i32 {
  return a;
}

fn b(a : ptr<private, i32>) -> i32 {}  // Current ptr syntax

var<private> z : i32 = 2;
fn main() -> void {
  ref y : i32 = z;
  var c : i32 = b(y);
}
```

Using `ref` is more concise, and provides a clearer idea of what you're getting. It functions the same as `var` in that we will auto-dereference to do the load/store as needed.

**Comment by dneto0** [comment]
A ref must have an initializer that would be an l-value in C.   It must be a reference to a variable or inside a variable.

Both var and ref always reference valid storage.  However, var declares *new* disjoint storage, and ref borrows/aliases existing storage.

We already have the [rule](https://gpuweb.github.io/gpuweb/wgsl.html#variable-declaration):

> A variable declaration ... Ensures the execution environment allocates storage for a value of the store type, for the lifetime of the variable.

We also have, for var:

> Two variables with overlapping lifetimes will not have overlapping storage.

But that does *not* apply to 'ref': the point of 'ref' is to point to already-existing storage in a variable.


Lastly, I would say that the storage class on a ref declaration is optional if it would be optional on a corresponding var declaration.  This would take effect for function arguments that are textures or samplers.

```
[[group(0),binding(0)]] var t : texture2d<f32>;
[[group(0),binding(1)]] var s : sampler;

fn getit(ref tparam : texture2d<f32>, ref sparam : sampler) -> vec4<f32> {
   return textureSample(tparam, sparam);
}
``` 
(I would argue that textures and samplers should only be passed around by reference. But maybe that's a separate wrinkle.)

**Comment by toji** [comment]
Having recently bumped into this myself I'm in favor of the simpler syntax, and the keyword `ref` reads to me as more suggestive of the usage. To sanity check, my assumption from the above is that you could use this syntax to reference structs as well, since I see that as being one of the more common uses of such syntax.

For example, this slightly more complicated example would work in theory, yes?
```rust
[[block]] struct Particle {
   [[offset(0)]] pos : vec2<f32>;
   [[offset(8)]] vel : vec2<f32>;
};
[[block]] struct ParticleBuffer {
   [[offset(0)]] particleCount : i32;
   [[offset(4)]] particles : [[stride(16)]] array<Particle>;
};
[[set(0), binding(0)]] var<storage> particleBuffer : [[access(read_write)]] ParticleBuffer ;

fn updateParticle(ref particle : Particle) -> void {
  particle.pos = particle.pos + particle.vel;
}

[[stage(compute)]]
fn main() -> void {
  // Ignore the terrible use of GPU workgroups, please! ;)
  for (var i :  i32 = 0; i < particleBuffer.particleCount; i = i + 1) {
    ref particle : Particle = particleBuffer.particles[i];
    updateParticle(particle);
  }
}
```

**Comment by kvark** [comment]
oh boy, now we are going to have `const`, `let`, `var`, `ref`, and we aren't done yet

**Comment by kvark** [comment]
Aren't we going to have pointers anyway at some point? If so, the use cases provided here can be fulfilled by using the pointers. There is no need to invent new constructs and grammar.

**Comment by dj2** [comment]
I don't think we've discussed pointers at all this point? The current `ptr` is confusing both from a usage and a communication perspective. Saying a `var` is a pointer to storage, but it's different from `ptr` is weird.

**Comment by kvark** [comment]
> Saying a var is a pointer to storage, but it's different from ptr is weird.

The thing I've been saying from the very start is that maybe we shouldn't have `var` as this special snowflake to start with. Remember #622 ? If load and stores are explicit, then you don't need a separation between const/var/ref, you can have one and only thing, say `let`:
```
let x: i32 = 3;
let y: pointer<private, i32> = &4;

fn foo(z: pointer<private, i32>) {
  *z = 5;
}
```

**Comment by dneto0** [comment]
Hey @toji:  your example has two problems:
- minor problem: you'd have to mention the storage class on the formal parameter to the function.  So  `fn updateParticle(ref<storage> particle : Particle)` 
- however, your example would not be allowed because base functionality in SPIR-V/Vulkan, and also HLSL does not support pointer-to-storagebuffer as function arguments.  In Vulkan you'd need an extra VariablePointers capability, which is future expansion potentially.

**Comment by dneto0** [comment]
Hey @kvark, you can't have just one concept, because you need something to actually declare that you have unique new storage (a variable), vs. the name for a computed value (what 'const' is today).

Or are you saying that `let y: pointer<private, i32> = &4` is the declaration of new storage?   In that case how do you form a pointer to existing storage.

I'd argue that #622 was about making loads/stores explicit.  But this issue is about what syntax to use for "I'm declaring a new named thing that points to the interior of storage in this other named thing".

**Comment by kvark** [comment]
@dneto0 that's a great point, I didn't see it clearly from this perspective.

> Or are you saying that let y: pointer<private, i32> = &4 is the declaration of new storage? In that case how do you form a pointer to existing storage.

It doesn't matter, strictly speaking - we can figure out how exactly to make it work. For example:
```rust
let x: pointer<function, vec2<i32>> = alloc(vec2<i32>(4)); // allocate new value in the "function" storage class
let y: pointer<function, i32> = x.x; // make a pointer to X component
let z: pointer<function, i32> = y; // make another pointer to the same value
```

The `alloc` would just be a regular function defined as:
```
fn alloc(value: T) -> pointer<function, T>;
```
The "." operator would be working just like "OpAccessChain": the argument is a pointer, and the result is another pointer.
Or we can make it work with a de-reference, these are details we don't have to resolve for choosing one path over the other.

> I'd argue that #622 was about making loads/stores explicit. But this issue is about what syntax to use for "I'm declaring a new named thing that points to the interior of storage in this other named thing".

I find these highly related. If you have explicit load/stores, then expressing the "points to the same thing" concept with pointers becomes natural. It's only when you hide load/stores that the reference concept becomes a problem.

To clarify, I'm not necessarily trying to say that we made a wrong decision in #622. I'm just pointing out that there is another path here, which is more explicit, maps to SPIR-V better, and it only needs `let` instead of `var/const/ref`. Just something to think about ;)

**Comment by dneto0** [comment]
Yes, I agree there is another path with an alloc-like builtin.  That path would also need a special rule about how lifetimes work, e.g. that alloc in function storage expires at the end of the scope, which is a bit magic. (It's exactly alloca in LLVM).  And you'd need a different rule for lifetimes in other storage classes.

I think we're having the right debate.   The "ref" idea grew out of folks making a bunch of samples, and their struggles with usability / conciseness / confusion.

I'd like to raise some other factors:
- we haven't talked much about function parameters of pointer type.  It's somewhat long-winded, just like other uses of the ptr syntax.
- Each value still has a type, and I feel strongly that we should have an unambiguous syntax to write down that type.  So even a ref would have a _type_ of `ptr<blah,store-type>`, even if that's not what you would commonly write in WGSL.

**Comment by toji** [comment]
> Hey @toji: your example has two problems:

Thanks for clarifying! A couple of questions to make sure I understand the limitation:
  
  - If the same struct was given to the shader as a uniform would it be valid to pass as a function param?
  - If the same logic was handled without a function (like below) it sounds like it would be allowed? 
   ```
  for (var i :  i32 = 0; i < particleBuffer.particleCount; i = i + 1) {
    ref<storage> particle : Particle = particleBuffer.particles[i];
    particle.pos = particle.pos + particle.vel;
  }
   ```

**Comment by dneto0** [comment]
> If the same struct was given to the shader as a uniform would it be valid to pass as a function param?

No, because of Vulkan/SPIR-V limitations: 
A GLSL uniform (not input and not output) is mapped to a uniform buffer object, which can't be passed by reference/pointer to a function.

SPIR-V / Vulkan allows function parameters that are pointer to:
 - input (pipeline inputs)
 - output (pipeline outputs)
 - workgroup storage class (a.k.a. "shared" in compute shaders)
 - "function" storage class: variables defined within a function body.
 - "private" storage class:  per-thread variables at module scope; in GLSL these are declared at module scope that are *not* inputs, outputs, shared, buffers, textures, or samplers.

(See [2.16.1 Universal Validation Rules](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_validation_a_validation_rules) in the part where it says "Any pointer operand to an OpFunctionCall must...".  Note that AtomicCounter is also listed there, but that's only for OpenGL environments.  In Vulkan atomics variables are in other storage classes.)

Now, it might be that we discover down the road that DXIL restricts function parameters somehow to limit us further.  At this point I don't see anything that would imply that

> If the same logic was handled without a function (like below) it sounds like it would be allowed?

Yes, that's allowed, because the reference/pointer is used in the same function where it's taken (originally defined).  That's the purpose of being able to take references: to do an address calculation early and reuse it many times within scope.

**Comment by dneto0** [comment]
I forgot: SPIR-V/Vulkan allows pointer-to-sampler and pointer-to-image as function parameters (They are in UniformConstant storage class).
See #1139  which tracks restrictions on function parameters

**Comment by jdashg** [comment]
Can (base) SPIR-V pointers be non-static but uniform? Or do they ultimately need to be staticly known at compile time?

**Comment by dneto0** [comment]
> Can (base) SPIR-V pointers be non-static but uniform? Or do they ultimately need to be staticly known at compile time?

In base SPIR-V (for graphics, which is what we're going for) pointers in general can be non-static and non-uniform.

For example, you can  have a loop that uses a newly computed pointer for each iteration, and then use it.

e.g. the equivalent of:

```rust
fn main() -> void {
      var arr : array<i32,10>;
      var i : i32 = 0; 
      loop {
          if (i32 >= 4) { break; }
          ref elemref : i32 = arr[i];   // compute a pointer to element i.     current wgsl would be
                // current wgsl would be:           const elemref : ptr<function,i32> = arr[i];
          elemref = 0; // start using it, updating the array element.
          elemref = elemref + 15;  // use it many times.
     }
}
```

However, base (graphics) SPIR-V restricts where pointers can go:
 - can't compute the select of two pointers  (e.g.        cond  ? ptrA : ptrB;  is not allowed)
 - can't return a pointer from a function
 - can't pass pointer to a storage buffer or uniform buffer into a helper function.
 - can't pass pointer to the _interior of_ a storage buffer or uniform buffer into a helper function.
 - can't store or load a pointer from a variable

In practice so-called logical pointers into storage buffer or uniform buffer are constrained enough that you can always trace back from a pointer value back to the originating variable declaration.  So effectively you statically know which variable.  But the interior location (e.g. index into an array) might not be uniform.

In base SPIR-V you *can* pass pointers into helper functions when they are pointer to other storage classes, e.g. Function, Private, or UniformConstant (a.k.a. images and samplers).

You can have arrays of textures, but in base SPIR-V, if you do a texturing (sampling) operation, I believe the array index has to be uniform across the invocation group (think quad) that is executing that specific operation.  There are extensions to allow non-uniformity of the index.   Note that I haven't concerned myself about that for WGSL because it doesn't have arrays of textures or samplers (yet).

Sorry for the long-winded answer.  It's got a bunch of rough edges.  I plan to make the type rules for pointer evaluation strict enough so that it ends up enforcing the base SPIR-V constraints.   Also I already filed an issue about the limitations for function parameters.

**Comment by dneto0** [comment]
I posted a PR for how pointers *currently* work. See #1368 
I'd like folks to absorb that before proceeding much further on this issue.  It has a bunch of relevant examples in the "Pointer types" section.

**Comment by jdashg** [comment]
I still think it's nicer to jump straight into ptr syntax:
```
for (var i :  i32 = 0; i < particleBuffer.particleCount; i = i + 1) {
 let particle = &(particleBuffer.particles[i]); // ptr<function, Particle>
 particle.pos = particle.pos + particle.vel;
}
```

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-01-26-Minutes#removing-ptr-and-adding-ref-1334">WGSL meeting minutes 2021-01-26</a></summary>

*   DM: Lot of discussion about ptr semantics that DN write. Some disagreement on how this should be rolled out. Outcome of that will change refs. 
*   MM: So, alloca
*   DN: I think alloca is a poor solution.
*   DM: Argue.
*   DN: Alloca is a function which returns a pointer which we can’t do in general so looks funny. Pointer implicitly escapes. Magical thing that happens to storage when you leave scope.
*   KN: alloca is dynamic stack allocation, variables are static stack allocation
*   MM: There is another argument because the alloca argument has to be a type and we have no concept of functions accepting types. This isn’t malloc as we need to be more type safe.
*   JG: Preference is the bare c-style. You declare a variable on stack and it gets memory. If you want to refer to it by pointer take the ref and use that. Existing rules for pointer validation. If you want to access, explicitly dereference and get the thing pointed too. I think that’s a useful way to surface what's happening under the hood for folks who are familiar with that. Lot of existing literature. Working in the constraints of something like c++ refs. Think it's OK to say we have different rules for when you can/can’t assign pointers and can’t do pointer math, that’s fine. Phrased in terms of c-style pointer syntax makes the most sense to me. Has the most prior art. Less confusing then what we currently have with var. Fear it’s more verbose and annoying.
*   DM: If you have syntax to declare vars on the stack you can take a pointer from, what can you do with them? Is it special syntax for special syntax? If you can only access a pointer why not hide behind initialization for the pointer? \
JG: If you combine into 1 step, get var today which I find unsatisfying.
*   DM: I meant an alloca function which we don’t have
*   JG: alloca has other problems since no-one knows how it works. Would be hesitant to use that and as KN said, dynamic not static. Try to emphasize if you want on stack put on stack, if you want to pass pointer to something on stack then use that
*   DM: Would like those objections written down on the issue somewhere. Feel like it’s easy to argue for alloca being the solution here.
*   JG: I think there is consensus alloca is not the way to go.
*   DN: I agree. Wrote previously that the lifetimes of an alloca look very magic and don’t gain anything out of it.
*   MM: Don’t think we have anything to lose by trying to type up the arguments more formally cand coming back a week from now.
*   DN: Lost on where we are, why are we talking about alloca and not ref? Ref is, I think, reasonable but I think it makes lifes easier. Saw DNs argues that there is another keyword but it’s another keyword for another concept. Also reduces typing.
*   MM: One question, in a world where we have real pointers (spir-v variable pointers) will we have `ref ref ref ref ref`.
*   DN: Spec says a var is a decl of storage for a name of a certain type. A ref is going to be a const declaration of a pointer value. The value of the pointer does not change. The type is a pointer type which still exists. Just a different way to declare a const of pointer type. Could have ptr -> ptr. Could have a ref to storage which holds a pointer which is a ref of a reference. The thing that exists is pointer types and pointer values. Ref is shorthand for declaring.
*   MM: Trying to think of the code I’d write if I wanted to make a ref to a variable and then the next line is what do I write when I want to deref.
*   DN: So, Brandon had a particle example. “18 days ago” comment had an example of ref into a storage buffer of type Particle then derefs a few times. That should work.
*   MM: So, there is no deref operator
*   DN: Right
*   MM: When he mentions Particle on the right, the type is Storage
*   DN: No, the type is ptr&lt;Storage, Particle>. Still ap pointer type
*   MM: Just implicitly deref’d
*   DN: Gets implicitly deref’d when used in context requiring a non-pointer result.
*   JG: Not favourite way to describe pointers to things. Where some of my concern comes from which is why it’s more a traditional approach. See this is easy to write and usable when you understand it, but hard to bolt onto existing understanding, have to learn this way on how it works. Harder then if we had a c-style, this is a pointer&lt;Storage, Particle> and it points to the thing.
*   MM: In a world with variable pointers, if I wanted a ref that references the variable named particle and I wanted to store through the outer ref to change where particle points too, is that possible?
*   DN: No, the ref declaration is a constant declaration for Particle.
*   MM: K, so with variable pointers, pointer can not be redirected to things
*   JG: We’d still have pointers, ref is just a shorthand. If you wanted to re-assign pointers you’d have to use the pointer syntax. Then when variable pointers could reassign that thing. We don’t today so can’t reassign but the pointer syntax would still exist.
*   DN: Might be mis-lead by the name variable pointers. In Vulkan that means other things than re-assignability.
*   MM: Ok.
*   DN: Thing that gets wrapped up is we don't have explicit deref of a pointer. There is no unary * from c. Maybe that is the thing causing weirdness and confusion
*   MM: If i added `ref particle2 = particle` it would not be a ref to a ref, I’d have two things pointing to the same storage.
*   DN: Correct. Which is like c++.
*   MM: Yes, it is. I C. Ok, figuring out how this works now. So before it would have been impossible to have ref of ref. That would be a validation rule, we’d just disallow it. Here, we’re fundamentally driving into the language by removing from the type system and replacing with a different token. You just can’t have ref of ref of ref. There is no way to type that in the language.
*   DN: C allows storing a pointer and copying it. WGSL design is that pointers are not storable. You can’t put them in memory and pass around. YOu can form a constant pointer value and point into the inside of something, and pass that around, but only as constants. Can’t store/modify it as data.
*   MM: So, if understanding, this is the same as `inout` in GLSL, but you can put this on a local var instead of just params?
*   DN: If you think of each statement as each statement as a function with `inout`, yes, but there are limitations on aliasing.
*   MM: Getting into Monads
*   DN: Your entire local store is a Monad.
*   MM: Totally.
*   DM: Is the need that refs cover, is it not covered by your pointers PR. \
DN: It is
*   DM: Coud have a pointer arg in a function and mutate the pointer
*   DN: Ref is a shorthand for the pointer type. So when DS makes code with a const and assigns to the const, it’s not that it’s constant it’s that it’s a fixed point in storage.
*   DM: In a function you aren’t typing const.
*   DN: True, not sure what the primary use case was, but same solution.
*   MM: One note, Swift has something similar. In swift, when declaring a class as `class foo{` but could say `struct foo{`. Would think they’re the same but they aren’t. Things as `class` are references. Things as `struct` are by value. There is at least one language with this kind of thing.
*   KN: C# also does that
*   MM: Warming up to this. Would like to discuss with RM.
*   DM: Feels unnecessary in the presence of pointers. Can postpone to post-MVP.
*   DN: True, could be pure syntactic sugar.
*   MM: Distinction between ref and ptr in wgsl is similar to c++. Don’t think devs struggle under the weight of those concepts
*   DM: They aren’t? Would disagree.
*   MM: With the distinction between them
*   DM: Especially mutable arguments, where you can’t see it’s being mutated when you call the function. Don’t think that's a good style.
*   KN: Chromium changed their style to allow mutable references for non-null parameters.
*    JG: Good place to stop, let’s get some more comments on the issue. We talked about earlier, if we linked or inlined the minutes onto a bug they’d be more surfacible
*   MM: In other groups, IRC bot that speaks github API.
*   DS: Should cleanup ahead of time as spelling is bad …
*   JG: Can clean it up. If these notes, and if you don't want something minuted ask for it not to be, should we post them?
*   MM: Inlining is good.
*   JG: Will try this week and see how it looks.
*   MM: Put inside a details element so it doesn’t get super long.
*   JG: Would like to be able to ctrl-f on the page. Hopefully that can see into a details element. Will try that, not too much work to do manually.
*   MM: Could also click on details elements and then cmd-f.

</details>

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-02-02-Minutes#removing-ptr-and-adding-ref--issue-1334--gpuwebgpuweb">WGSL meeting minutes 2021-02-02</a></summary>

*   (MM: Would like to discuss with RM)
*    MM: We discussed some concerns that when we add real pointers there will be 2 similar but distinct kinds of references and that’s distasteful. Mechanically the proposal would work, however. I assume that if we don’t take this proposal and we want ptr-to-ptr in the future, we would be using the same pointer type as today. Wouldn’t be new types; just rule relaxation.
*   DN: Would say yes. Current PR on how ptrs work has a rule that DM dislikes that if you mention a pointer and use it in a context that takes a non-pointer you do a load. With ptr-to-ptr that rule is no longer useful.
*   MM: We’d have to add a new operator or something
*   DM: LIke a deref operator which we could have now and solve the problem?
*   DN: Way other PR works is that there is no lvalue or rvalue but mentioning the var means you have a pointer. If you want fewer rules do what’s in the PR. With deref there is ambiguity if you’re using value storing or address. In c have & for address-of. Have to reach to outer level to determine if doing calcs or address calcs.
*   DM: We know types we do ops on. If a user is adding pointers we error out, is that an issue?
*   JG: Right now, pointers implicitly decay to refs. If you have a + 3 today if a is a pointer what that does is loads a and adds 3 and gives a result. The pointers we have today are refs that way which is weird. Would definitely prefer to have it explicit as it resolves ambiguity.
*   DM: I see it as we don’t have pointers today. The section is incomplete and no-one is using them. Thought it was a good idea in the future. Don’t need to base off something today
*   DN: Don’t understand that, currently exactly as SPIR-V and LLVM handle them.
*   DM: No spirv has explicit deref and you’re making it implicit.
*   KN: Would be nice to have deref for pointes but requiring deref on var because var is technically a pointer isn’t great
*   MM: Similar point, previously we started we don’t want explicit load/store and think that’s an approximation. We want code to be readable and having loads/stores would look bad; doesn’t mean there isn’t a middle ground.
*   DM: Leave var as is, not confusing, works like any language. Not super happy but it works. If we have real pointers which aren’t implicitly deref solves the problem
*   JG: Would be happy to write a proposal to have explicit load/store for pointers.
*   DN: PR from last week has all type rules for assignment and type rules for pointers. If you have a mirror to that of ‘this is what it looks like’ and ‘this is why the type rules aren’t wild’ and you haven’t introduced lvalue and rvalue.
*   DM: I think lvalues are already there. Pointers don’t change anything. When parsing `struct.member` you already have to know it’s an lvalue if on the left side of assignment. If you see deref on the left side you know it’s an lvalue. 
*   DN: For me, lvalue and rvalue is mysterious for people reading code. Can see a phrase in language and know the type. Don’t know if you have a variable which on the outside takes the address of, don’t know if calculating. Even sizeof in c, not actually doing work. As programmers we don’t put complicated expressions on the inside to get fooled. Would like to see JG's proposal.
*   MM: In JGs proposal, if we had a local variable (not let, var) and add 4 and assign to another local, no pointers involved, would that involve pointers and derefs and address of?
*   JG: No, would just work like for a normal variable. In some ways, var just says you handle the storage for this . Only time ptrs are involved is once you start using the reference of op to get the pointer to something. Now you have a pointer and there is no implicit decay. Really obvious when you have a pointer and it behaves differently.
*   DN: One thing to be clear. Later when we have a Memory Model, we have to look at knowing when memory accesses occur. Not an issue for private but in shaders we need to know when they occur and what statements. Probably Ok.
*   MM: Not sure it was stated, what the distaste of having lvalues and rvalues is?
*   DN: It’s hard to understand when you have an lvalue context vs rvalue context.
*   JG: Because you have to walk to top of expression tree
*   DN: Maybe that melts away as no pointer return from functions so anything on left is lvalue and anything on right is rvalue. If someone does that analysis and says it’s that simple I’d be happier. If the 5 node lattice in the c++ spec, trying to explain is a mess.
*   JG: Will move to the needs action column. Weird we have interdependent issues but don’t have meta issues. Weird that PRs are issues.
</details>

**Comment by jdashg** [comment]
I'm sympathetic to s/ptr/ref/ maybe, and that's where some of my proposals ended up.

However, making it a keyword instead is too far for me in terms of added complexity.
I would prefer it remained part of the for-reals type system, instead of a (keyword,type) tuple, which I think we should move away from in general.

Ideally we have let and const for our SSA id equivalents, and everything else is typed through them.

Also I think that we absolutely need to s/var/somethingelse/ because it's really confusing to have "variables" that may or may not be Variables.

Part of the thrust of this issue's proposal is to help reduce verbosity too, but I think that's best served by type inference.

Here's what I've been thinking:
```rust
fn get_or_init(a: ref<in, i32>, c: i32) -> i32 {
  if (a == 0) {
    a = c;
  }
  return a;
}

const x = decl_storage<in, i32>(); // -> ref<in, i32>
// Formerly `var<in> x : i32;`
const y = decl_storage<in, i32>(); // -> ref<in, i32>
fn main() -> void {
  const effective_y = get_or_init(y, x);
  [...]
}
```

```rust
for (let i : i32 = 0; i < *particleBuffer.particleCount; i = i + 1) {
 const particle = particleBuffer.particles[i];
 particle.pos = particle.pos + particle.vel;
}
```
You might notice there that it looks like we're modifying a const's members, and we are. This is how `const` works in JS, so that's a win for consistency with our main API binding language.

Alternatively with ptr and explicit deref:

```rust
fn get_or_init(a: ptr<in, i32>, c: i32) -> i32 {
  if (*a == 0) {
    *a = c;
  }
  return *a;
}

const x = decl_storage<in, i32>(); // -> ptr<in, i32>
const y = decl_storage<in, i32>(); // -> ptr<in, i32>
fn main() -> void {
  const effective_y = get_or_init(y, *x);
  [...]
}
```

```rust
for (let i : i32 = 0; i < *particleBuffer.particleCount; i = i + 1) {
 const particle = particleBuffer.particles[i]; // Subaccess(ptr<T>,K) -> ptr<typeof(T.K)>
 *particle.pos = *particle.pos + *particle.vel;
}
```

What's *really cool* about this is that all the load/store sites are obvious! "Wow I keep having to type star a lot" naturally pushes people to use intermediaries, which is what they should be doing anyway!

**Comment by jdashg** [comment]
I believe this is closed by #1569.
